Allocation module documentation (#5062)
wazuh/wazuh-qa: deployability/modules/allocation/README.MD

## Allocation Module

### User documentation

> The Allocation module allows you to create and destroy VMs both locally and in AWS, it also gives the possibility of creating macOS VMs on Wazuh infrastructure or PowerPC VMs, as long as you have the necessary permissions to do so. VMs can be AMD64, ARM64, Windows, macOS (Intel and ARM), and powerPC (Centos and Debian).

Reviewer: ```suggestion
The Allocation module allows you to create and destroy VMs both locally and in AWS, it also gives the possibility of creating macOS VMs on Wazuh infrastructure or PowerPC VMs, as long as you have the necessary permissions to do so. VMs can be AMD64, ARM64, Windows, macOS (Intel and ARM), and PowerPC (Centos and Debian).
```

> The Allocation module allows you to create and destroy VMs both locally and in AWS, it also gives the possibility of creating macOS VMs on Wazuh infrastructure or PowerPC VMs, as long as you have the necessary permissions to do so. VMs can be AMD64, ARM64, Windows, macOS (Intel and ARM), and PowerPC (Centos and Debian).

Reviewer 2: Does the current module deploy all those systems and architectures?

Also, Centos -> CentOS

Reviewer 3: Done a81aec813

#### Set the environment

> The execution of the allocation is carried out through the Workflow library, or by executing them manually through commands.

Reviewer 2: Workflow engine library

Reviewer 3: Done a81aec813

Execution can be done from any operating system.

Initially, You must install Python libraries. We recommend to use virtual environments. Follow the technical documentation at https://docs.python.org/3/library/venv.html.

1. Activate the environment:

  ```bash
  source {venv directory}/bin/activate
  ```

2. Clone the `wazuh-qa` repository:

  Navigate to the project directory and switch to the project branch:

  ```bash
  cd wazuh-qa
  git checkout {project-branch}

>   ```

Reviewer 2: Add a git clone of wazuh-qa using the HTTPS reference please

Reviewer 3: Done a81aec813

> Note: temporary dev project-branch is [enhancement/4495-DTT1](https://github.com/wazuh/wazuh-qa/tree/enhancement/4495-DTT1)

3. Install requirements:

  ```bash
  pip3 install -r deployability/deps/requirements.txt
  ```

#### Use the Allocation module through the Workflow

> Now, it is possible to use the worklow engine library to launch the provision module by doing the following steps:

Reviewer 2: Workflow engine

Reviewer 3: Done a81aec813

1. Install the Workflow engine library and its launcher:

  While in wazuh-qa:

  ```bash
  cd modules
  pip3 uninstall -y workflow_engine && pip3 install .
  ```

2. Test Fixture to Execute:

      It will be necessary to create a fixture (yaml file) where the infrastructure, provisioning, and tests to be executed will be declared.

      >Note: It is possible to find some fixture examples in [deployability/modules/workflow_engine/examples/](../workflow_engine/examples)

      Example:

      ```bash
      version: 0.1

>       description: This workflow is used to test agents deployment por DDT1 PoC

Reviewer: ```suggestion
      description: This workflow is used to test agents' deployment for DDT1 PoC
```

>       variables:

Reviewer 2: I don't know if we should add an example using other modules, or at least, make focus on the Allocation module more than the testing module

Reviewer 3: Done ee033425e

              args:
                - modules/allocation/main.py
                - action: delete
                - track-output: "{working-dir}/agent-{agent}/track.yaml"
          foreach:
            - variable: agents-os
              as: agent
      ```

      Following the schema of the example:

      Configure the following parameters depending on your test case:

      ```yaml
      variables/agent-os
      variables/manager-os
      infra-provider
      working-dir
      tasks
      ```

      Pay attention to the tasks:

      ```yaml
      args
      depends-on
      ```

      >Note: In args, configure the launcher's path correctly (main.py files in each module), and to fill `depends-on`, consider the steps of your test (allocation, provision, and test)

3. Execution of Command (local):

  Execute the command by referencing the parameters required by the library (launcher).

  ```bash
  python3 -m workflow_engine {.yaml fixture path}
  ```

  Example

  ```bash
  python3 -m workflow_engine modules/workflow_engine/examples/dtt1-agents-poc.yaml
  ```

  > Note The command execution can also be mediated through Jenkins.

#### Manual execution of the Allocation module

> If one wishes to execute the allocaation module without installing the workflow engine, they can proceed by using the launcher ([module/allocation/main.py](main.py)):

Reviewer: ```suggestion
If one wishes to execute the allocation module without installing the workflow engine, they can proceed by using the launcher ([module/allocation/main.py](main.py)):
```

> If one wishes to execute the allocation module without installing the workflow engine, they can proceed by using the launcher ([module/allocation/main.py](main.py)):

Reviewer 2: Workflow engine

Reviewer 3: Done a81aec813

1. Create

  While in wazuh-qa/deployability

- Local deployment (Vagrant)

  ```bash
  python3 modules/allocation/main.py --action create --provider '{{ vagrant }}' --size '{{ large }}' --composite-name '{{ composite-name }}' --inventory-output '{{ inventory }}' --track-output '{{ track }}'

  ```

  Example:
  ```bash
  python3 modules/allocation/main.py --action create --provider vagrant --size large --composite-name linux-ubuntu-22.04-amd64 --inventory-output "/tmp/dtt1-poc/agent-linux-ubuntu-22.04-amd64/inventory.yaml" --track-output "/tmp/dtt1-poc/agent-linux-ubuntu-22.04-amd64/track.yaml"
  ```

- AWS deployment

  ```bash
  python3 modules/allocation/main.py --action create --provider '{{ aws }}' --size '{{ large }}' --composite-name '{{ composite-name }}' --inventory-output '{{ inventory }}' --track-output '{{ track }}' --label-termination-date '{{ termination-date }}' --label-team  '{{ team }}'

  ```

>   >Note: In the case of AWS it is mandatory to define 2 arguments that are not necessary for Vagrant, --label-termination-date and --label-team.

Reviewer 2: Avoid using numbers, refer to them by their names: `two`, and apply this to the entire document

Reviewer 3: Done a81aec813

>    --label-termination-date: This argument allows you to define the date on which the machine can be deleted. The allowed values are 1d (where the 1 refers to the number of days the machine is needed) or with the following format "2024-03-20 21:00:00"

Reviewer 2: Add emphasis to 1, we need to make sure that we are referring to the value, and apply this to the entire document

> are `1d` (where the `1` refers

Reviewer 2: > The allowed values are **1d** (where the **1** refers to the number of days the machine is needed) or with the following format "2024-03-20 21:00:00"

or something like

> The allowed values are '1d' (where the '1' refers to the number of days the machine is needed) or with the following format "2024-03-20 21:00:00"

Reviewer 3: Done a81aec813

>    --label-team: This argument allows you to set the team that owns the VM to be able to track it. The valid options are: 'qa', 'core', 'framework', 'devops', 'frontend', 'operations', 'cloud', 'threat-intel', 'marketing', 'documentation'

Reviewer 2: `documentation` is `qa` right now

  Example:
  ```bash
  python3 modules/allocation/main.py --action create --provider aws --size large --composite-name linux-ubuntu-22.04-amd64 --inventory-output "/tmp/dtt1-poc/agent-linux-ubuntu-22.04-amd64/inventory.yaml" --track-output "/tmp/dtt1-poc/agent-linux-ubuntu-22.04-amd64/track.yaml" --label-termination-date "2024-03-20 21:00:00"  --label-team devops
  ```

2. Delete

  While in wazuh-qa/deployability

  ```bash
  python3 modules/allocation/main.py --action delete --track-output '{{ track }}'

  ```

  Example:
  ```bash
  python3 modules/allocation/main.py --action delete --track-output "/tmp/dtt1-poc/agent-linux-ubuntu-22.04-amd64/track.yaml"
  ```

  >Note: The --track-output argument is mandatory for the delete action because this file contains all information of the VM that will be destroyed

3. Arguments

  - --provider

>     This argument allows us to choose on which platform we will deploy our VM. The allowed values are aws or vagrant.

Reviewer 2: Same, please emphasize the allowed values, at first I thought they were the service/software names (apply this to all the document)

Reviewer 3: Done a81aec813

    >Note: this argument is mandatory for delete action.

  - --inventory-output
    This argument allows us to define which path we want the inventory file to write to. By default, it is saved in the instance directory, for example: /tmp/wazuh-qa/VAGRANT-F0753C57-713E-4294-9EA9-89D93D384844/inventory.yml

  - --working-dir
    This argument allows us to define in which directory the files referring to the VM will be generated. By default, /tmp/wazuh-qa

  - --label-issue
    This argument is only used in the case of AWS and is not mandatory, it allows us to create a label to reference the created instance to an issue on GitHub. It has to be a GitHub URL of the Wazuh repository, for example: https://github.com/wazuh/internal-devel-requests/issues/1008

  - --label-team
    This argument it is mandatory for AWS deploy, allows you to set the team that owns the VM to be able to track it. The valid options are: 'qa', 'core', 'framework', 'devops', 'frontend', 'operations', 'cloud', 'threat-intel', 'marketing', 'documentation'

  - --label-termination-date

>     This argument it is mandatory for AWS deploy, allows you to define the date on which the machine can be deleted. The allowed values are 1d (where the 1 refers to the number of days the machine is needed) or with the following format "2024-03-20 21:00:00"

Reviewer: ```suggestion
    This argument is mandatory for AWS deployment, it allows you to define the date on which the machine can be deleted. The allowed values are 1d (where the 1 refers to the number of days the machine is needed) or with the following format "2024-03-20 21:00:00"
```

  - --instance-name
    This argument allows us to define a custom name for the instance, if this argument is not used, the instance name is defined by other parameters entered, such as --label-issue or --composite-name.
---

### Technical documentation

The allocation module allows creating infrastructure on both AWS and locally (using Vagrant).

> Instructions can be initiated from the fixture and executed through the Workflow Engine or executed using Python commands.

Reviewer 2: Please change to `Workflow engine`

Reviewer 3: Done a81aec813
